url27db8982cd_leohinterlang_argFace_tgz-pJ8-com_fidelis_argface_ArgSampleJ8/src/com/fidelis/argface/ArgPattern.java

```diff
@@ -0,0 +1,633 @@
+        }
+        patternList.add(node);
+        if (ender) {
+            sb.append('x');
+        } else {
+            sb.append('L');
+            adjustCount(node);
+        }
+        ++prevIndex;
+    }
+    
+    private void buildReduce (ArgNode node, StringBuilder sb) {
+        ArgOperand operand = new ArgOperand().literal();
+        operand.setName("$END$");
+        ArgNode reduce = new ArgNode(operand);
+        patternList.add(reduce);
+        sb.append('x');
+        ++prevIndex;
+    }
+    
+    private void buildVariable (ArgNode node, StringBuilder sb) {
+        patternList.add(node);
+        sb.append('V');
+        adjustCount(node);
+        ++prevIndex;
+    }
+    
+    private void adjustCount (ArgNode node) {
+    	++patternMin;
+        if (node.isRepeat()) {
+            patternMax = 999;
+        }
+    }
+    
+    private String buildGroup (ArgNode node, StringBuilder sb, boolean rebuild) {
+        patternList.add(node);
+        sb.append('g');
+        ++prevIndex;
+        ArgList group = node.getGroup();
+        listStack.push(argList);
+        argList = group;
+        String pattern = null;
+        if (rebuild) {
+            pattern = rebuildPattern(argList.goHome(), sb);
+        } else {
+            pattern = buildPattern(argList.goHome(), sb);
+        }
+        argList = listStack.pop();
+        return pattern;
+    }
+
+    private char remLast (StringBuilder sb) {
+        int last = sb.length() - 1;
+        if (last < 0) {
+            return '\0';
+        }
+        char c = sb.charAt(last);
+        sb.setLength(last);
+        return c;
+    }
+
+    private String patternSpec (String pattern) {
+        StringBuilder sb = new StringBuilder();
+        ArgOperand operand = null;
+        int n = 0;
+        for (ArgNode node : patternList) {
+            char c = pattern.charAt(n++);
+            boolean trailers = false;
+            if (c == 'L') {
+                if (node.isOptional()) {
+                    sb.append('[');
+                }
+                operand = node.getOperand();
+                String literal = operand.getName();
+                sb.append(literal);
+                trailers = true;
+            } else if (c == 'V') {
+                if (node.isOptional()) {
+                    sb.append('[');
+                }
+                operand = node.getOperand();
+                String name = operand.getName();
+                sb.append('<');
+                sb.append(name);
+                sb.append('>');
+                trailers = true;
+            }
+            if (trailers) {
+                if (node.isOptional()) {
+                    sb.append(']');
+                }
+                if (node.isRepeat()) {
+                    sb.append("...");
+                }
+                sb.append(' ');
+            }
+        }
+        remLast(sb);
+        return sb.toString();
+    }
+    
+    private void addTargetOperands (String pattern) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.